src/compiler/checker.ts

```diff
@@ -6170,6 +6174,37 @@ namespace ts {
             }
         }
 
+        function containsArgumentsReference(declaration: FunctionLikeDeclaration): boolean {
+            const links = getNodeLinks(declaration);
+            if (links.containsArgumentsReference === undefined) {
+                if (links.flags & NodeCheckFlags.CaptureArguments) {

```

Comment:
I will clarify. When we resolve a call, we do not necessarily check the body of the function. As a result, we must traverse the body of the function to discover whether `arguments` is used in an expression. *As an optimization*, we can use the pre-existing `CaptureArguments` flag if the function was checked *prior* to resolving the call. We cannot guarantee that `CaptureArguments` has already been set.